app/src/kalman_soc.cpp

```diff
@@ -0,0 +1,399 @@
+/*
+ * TinyEKF: Extended Kalman Filter for embedded processors
+ *
+ * Copyright (C) 2015 Simon D. Levy
+ *
+ * MIT License
+ */
+
+#include "kalman_soc.h"
+#include <math.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#define DEBUG
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(kalman_soc, CONFIG_LOG_DEFAULT_LEVEL);
+
+typedef struct
+{
+
+    float *x; /* state vector */
+
+    float *P; /* prediction error covariance */
+    float *Q; /* process noise covariance */
+    float *R; /* measurement error covariance */
+
+    float *G; /* Kalman gain; a.k.a. K */
+
+    float *F; /* Jacobian of process model */
+    float *H; /* Jacobian of measurement model */
+
+    float *Ht; /* transpose of measurement Jacobian */
+    float *Ft; /* transpose of process Jacobian */
+    float *Pp; /* P, post-prediction, pre-update */
+
+    float *fx; /* output of user defined f() state-transition function */
+    float *hx; /* output of user defined h() measurement function */
+
+    /* temporary storage */
+    float *tmp0;
+    float *tmp1;
+    float *tmp2;
+    float *tmp3;
+    float *tmp4;
+    float *tmp5;
+
+} EKF;
```

Comment:
Ah, this looks very similar to the struct in `bat_charger.h`. Why do we need it two times?

Reply:
I guess, the purpose is to have access to the data from other .c files, as we cannot use classes in C code. There is high provability that there are much easier solutions. Let's discuss it in a call as suggest in:  https://github.com/LibreSolar/charge-controller-firmware/pull/130/files#r866892157